[CAMEL-4271]  JDBCAggregateRepository and Recovery in a Cluster

bonnetb opened the pull request:
When a JDBCAggregateRepository is deployed in a cluster, recovery tasks from the different members may:
*  race with each other to recover completed exchanges that must be recovered
*  race with the normal process of an in progress completed exchange that was completed by another member of the cluster (when an exchange was completed by another member, recovery task is not aware that that exchange might still be in progress...) 

To solve that issue, we propose that each member is responsible only for the recovery of the exchange it completed. We add an instance_id column in the completed table in order to know who completed the exchanges.

A side effect is that if a member is stopped, its completed exchanges shall not be recovered until that member is restarted (or until an UPDATE of instance_id assign completed exchanges to another running member)

davsclaus commented:
Sorry this is not accepted to force that each node must only recover exchanges that came from this node. 

Instead you can make an extended class that has this feature and leave the original class as-is. Then add a section in the docs about this and that users can choose to use yours instead with this limitation.

bonnetb commented:
Back to main JdbcAggretationRepository, changing field visibility and I extended it in ClusteredJdbcAggregationRepository, updated doc, and added a JUnit test.
